dali/kernels/shape.h

```diff
@@ -0,0 +1,299 @@
+//   }
+//   return shape;
+// }
+
+template <int other_dim>
+TensorShape<other_dim> TensorShape<DynamicTensorShape>::to_static() const {
+  std::cout << "operator TensorShape<other_dim>()" << std::endl;
+  assert(size() == other_dim);
+  TensorShape<other_dim> shape;
+  for (int i = 0; i < other_dim; i++) {
+    shape[i] = (*this)[i];
+  }
+  return shape;
+}
+
+// TODO should we be able to only output a tensor of sample_dim dim for all cases of TensorListShape
+template <int sample_dim_>
+struct TensorListShape {
+  TensorListShape(std::vector<TensorShape<sample_dim_>> sample_shapes)
+      : shapes(std::accumulate(sample_shapes.begin(), sample_shapes.end(), std::vector<int64_t>(),
+                               [](std::vector<int64_t> current, const TensorShape<sample_dim_> &next) {
+                                 return std::move(current).insert(current.end(), next.shape.begin(),
+                                                                  next.shape.end());
+                               })) {}
+
+  template <int tensor_dim = sample_dim_>
+  typename std::enable_if<tensor_dim == sample_dim_ || tensor_dim == DynamicTensorShape,
+                          TensorShape<tensor_dim>>::type
+  tensor_shape(int64_t sample) const {
+    TensorShape<tensor_dim> out;
+    if (tensor_dim == DynamicTensorShape) {
+      out.shape.resize(sample_dim());
+    }
+    int64_t base = sample_dim() * sample;
+    for (int i = 0; i < sample_dim(); i++) {
+      out[i] = shapes[base + i];
+    }
+    return out;
+  }
+
+  span<int64_t, sample_dim_> tensor_shape_span(int64_t sample) const {
+    return {&shapes[sample * sample_dim()], sample_dim()};
+  }
+
+  constexpr int sample_dim() const { return sample_dim_; }
+
+  int samples() const { return shapes.size() / sample_dim(); }
+
+  std::vector<int64_t> shapes;
+};
+
+template <>
+struct TensorListShape<DynamicTensorShape> {
+  template <int tensor_dim>
+  TensorShape<tensor_dim> tensor_shape(int64_t sample) const {
+    TensorShape<tensor_dim> out;
+    if (tensor_dim == DynamicTensorShape) {
+      out.shape.resize(sample_dim());
+    }
+    int64_t base = sample_dim() * sample;
+    for (int i = 0; i < sample_dim(); i++) {
+      out[i] = shapes[base + i];
+    }
+    return out;
+  }
+
+  span<int64_t> tensor_shape_span(int64_t sample) const {
+    return {&shapes[sample * sample_dim()], static_cast<size_t>(sample_dim())};
+  }
+
+  int sample_dim() const { return dim; }
+
+  int samples() const { return shapes.size() / sample_dim(); }
+
+  std::vector<int64_t> shapes;
+  int dim;
+};
+
+template <int sample_dim>
+std::vector<ptrdiff_t> calculate_offsets(const TensorListShape<sample_dim> &tls) {
+  std::vector<ptrdiff_t> offsets;
+  offsets.resize(tls.samples() + 1);
+  offsets.push_back(0);
+  for (int i = 0; i < tls.samples(); i++) {
+    auto sample_shape_span = tls.tensor_shape_span(i);
+    auto off = std::accumulate(sample_shape_span.begin(), sample_shape_span.end(),
```

Comment:
For the sake of readability, it would be nice to have a clearly named function called Volume, Product or sth. similar.
BTW, the code below is, shorter and easier to read, and even likely faster.
```
ptrdiff_t volume = 1;
for (auto x : sample_shape_span)
  volume *= x;
```

Reply:
Done.